bin/dasht-docsets-inherit

```diff
@@ -0,0 +1,72 @@
+#!/bin/sh -e
+#
+# # DASHT-DOCSETS-INHERIT 1       2020-05-16                            2.4.0
+#
+# ## NAME
+#
+# dasht-docsets-inherit - unpack docsets from Dash.app
+#
+# ## SYNOPSIS
+#
+# `dasht-docsets-inherit` [*NAME*...]
+#
+# ### Examples
+#
+# `dasht-docsets-inherit`
+#   Unpack all installed Dash.app docsets.
+#
+# `dasht-docsets-inherit` sh
+#   Unpack Dash.app docsets whose names contain "sh".
+#
+# `dasht-docsets` sh 'c$'
+#   Unpack Dash.app docsets whose names contain "sh" or end in "c".
+#
+# ## DESCRIPTION
+#
+# Unpack specified Dash.app docsets. This is necessary to use the Docsets with
+# `dasht`. Docsets are specified by passing one or more name filters as
+# arguments, which are matched against docset names. If no filters are
+# specified, all docsets are unpacked. Note that `DASHT_USE_DASH_APP` must be
+# set to `true` for this script to execute.
+#
+# ## ENVIRONMENT
+#
+# `DASHT_USE_DASH_APP`
+#   Set to "true" to use Dash.app's (MacOS only) docset library. Disables
+#   docset management commands (install, remove, update). Also changes the
+#   default `DASHT_DOCSETS_DIR` to "~/Library/Application Support/Dash".
+#
+# ## EXIT STATUS
+#
+# 44
+#   No results were found.
+# 46
+#   Command was not executed because `DASHT_USE_DASH_APP` is not set.
+#  
+# ## SEE ALSO
+#
+# dasht-docsets(1), [Dash]
+#
+# [Dash]: https://kapeli.com/dash
+#
+# ## AUTHOR
+#
+# Written in 2016 by Suraj N. Kurapati <https://github.com/sunaku/dasht>
+# Distributed under the terms of the ISC license (refer to README file).
+
+: ${DASHT_USE_DASH_APP:=false}
+if $DASHT_USE_DASH_APP; then
+  : ${DASHT_DOCSETS_DIR:="$HOME/Library/Application Support/Dash"}
+else
+  exit 46
+fi
+
+# Ensure any docsets we are accessing from the Dash.app library have had their
+# documents extracted. Dash.app leaves the documents in the tarball by default.
+dasht-docsets "$@" | while read docset; do
+  for root in "$DASHT_DOCSETS_DIR"/*/*/"$docset".docset; do
```

Comment:
I think `$DASHT_DOCSETS_DIR` shouldn't be redefined, because the user may have set it to something special.  Instead, the loop should try looking in `$HOME/Library/Application Support/Dash` (which can be factored out as follows).

```sh
KAPELI_DASH_DIR="$HOME/Library/Application Support/Dash"
# ...
if test -e "$KAPELI_DASH_DIR"; then
  for root in "$KAPELI_DASH_DIR"/*/*/"$docset".docset; do
    # ...
```

Reply:
I think it's probably better for you to make the changes, because it seems like we have somewhat different architectural ideas about the interface, and I think if I made just the changes you requested, it would leave the project with a weird mixture-- if you want to move in the requested direction, you'll probably need to adjust some of my other changes as well. So I'll explain the overall design I followed, then you can do what you want with it.

When making my changes, this was my thinking: the user is using `dasht` in one of two "modes": regular mode or Dash.app mode. Dash.app mode is activated by setting the `DASHT_USE_DASH_APP` flag. All of the scripts behave conditionally on this flag-- that is why `dasht-docsets-inherit` does as well, it would be weird to have only script that didn't.

Similarly, `dasht-docsets-inherit` just follows the behavior of the other scripts re `DASHT_DOCSETS_DIR` as well-- it sets a default value, the macOS location (so a user setting would not be changed). In general, I don't expect people to simultaneously set `DASHT_DOCSETS_DIR` and `DASHT_USE_DASH_APP`-- the exception is if Dash.app is in some exotic location.

I was also a bit confused by the name `dasht-docsets-inherit` that you requested. To me, this suggests a kind of transferring of Dash.app docsets into `dasht`, e.g. Dash.app docsets would be copied over and extracted into the dedicated `DASHT_DOCSETS_DIR` directory. If that were the case, I would understand why you wouldn't want this called in the main `dasht` executable-- it would be a kind of separate import operation. But this is not how it works at present. All `dasht-docsets-inherit` does is perform in-place extraction of `Dash.app` docsets within their existing directory. This is a one-time operation that is necessary when accessing Dash.app docsets. Given the current design, I don't understand the rationale for running this separately. As it is, if the user downloads some datasets in Dash.app, they are seamlessly available in `dasht`. But if you make them run `dasht-docsets-inherit` separately, then every time they download a new Dash docset, they have to remember to run `dasht-docsets-inherit`, for no benefit that I can see.

It seems to me that your vision is better suited to a system where, rather than having `Dash.app` mode and regular mode (as I designed it), `dasht` always just reads from it's own dedicated docset database/directory, and you can import into this directory from either Dash.app or Zeal/custom whatever. That sounds like a good design, but that implies bigger changes (removing `DASHT_USE_DASH_APP` entirely).

Reply:
Well said! :100: You've understood my vision correctly and, thanks to your clear explanation, I now realize that I didn't quite understand your approach in this PR until now. :sweat_smile:  Let me think more about how to reconcile our different approaches. :thinking:

One of the reasons for my knowledge gap is that I don't have access to (and have never actually used) the [Dash for macOS](https://kapeli.com/dash) app firsthand.  So I don't know how Dash organizes its docsets on the filesystem nowadays, other than from a helpful (but now possibly outdated) user-provided tip in https://github.com/sunaku/dasht/issues/15#issue-161019412.

In particular, since you intended to extract Dash's docsets directly within its own internal area (as opposed to dasht's `DASHT_DOCSETS_DIR` area), I'm wondering how Dash operates: :thinking: it either has its own internal area where it extracts docset tarballs (in which case, couldn't we just point to that instead of re-extracting for dasht?) or it has some kind of virtual filesystem adapter that dynamically exposes the contents of docset tarballs as if they were normal extracted files on disk.

I'm intrigued by the latter possibility :nerd_face: and am curious to try implementing a very basic, shell scripted version thereof.  A direct consequence of this would be that `dasht` would need to become web-server based by default, since direct file access can't be intercepted by mere shell scripts: that would require a **_real_** VFS, which is beyond the scope of this project.

Could you help me understand how Dash organizes its docsets on your system?  A simple diagram, such as the output of the `tree` command, as exemplified in https://github.com/sunaku/dasht/issues/15#issue-161019412), would be sufficient.

Reply:
Glad I could clear things up!

Since the initial work on this was done close to a year ago, I don't remember everything I learned about Dash.app's internal organization of docsets when I started work on this, but I remember being surprised that I couldn't find extracted versions of the docsets in Dash's saved data. I don't know how it accesses the tarball content, I assumed it uses some library I was unaware of for reading tarballs directly, and since this was a shell script project, I didn't investigate further.

Here's an example of the structure of a DocSet as Dash stores it:

```
── SciPy.docset
    └── Contents
        ├── Info.plist
        └── Resources
            ├── docSet.dsidx
            ├── optimizedIndex.dsidx
            ├── tarix.tgz
            └── tarixIndex.db
```

And attached is the full output of `~/Library/Application Support/Dash`.
[tree.txt](https://github.com/sunaku/dasht/files/6874576/tree.txt)

Reply:
Thanks for the detailed report! :bow: There seems to be a mixture of fully extracted docsets (which is what dasht currently uses) as well as unextracted `tarix.tgz` ones.

Could you inspect the `tarixIndex.db` file further?  I'm thinking that it's either a repackaged version of `tarix.tgz` (to avoid the overhead of gunzip before accessing the contents of the underlying tar) or a searchable index thereof (to avoid the same gzip overhead of traversing the underlying tar). :thinking:

Reply:
That mixture is there because I use my fork of `dasht`, which does the extraction in place on-demand. I haven't searched every docset in my Dash library with `dasht`, so not every docset is extracted.

Here's a link to a tarixIndex.db file, it's a SQLite database (Github won't let me attach this file type directly to the issue). https://file.io/1Lsva6fX1MLA